OrbitProducer/CaptureEventProducer.cpp

```diff
@@ -0,0 +1,203 @@
+// Copyright (c) 2020 The Orbit Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "OrbitProducer/CaptureEventProducer.h"
+
+#include "OrbitBase/Logging.h"
+
+using orbit_grpc_protos::ProducerSideService;
+using orbit_grpc_protos::ReceiveCommandsAndSendEventsRequest;
+using orbit_grpc_protos::ReceiveCommandsAndSendEventsResponse;
+
+namespace orbit_producer {
+
+CaptureEventProducer::~CaptureEventProducer() = default;
+
+void CaptureEventProducer::BuildAndStart(const std::shared_ptr<grpc::Channel>& channel) {
+  CHECK(channel != nullptr);
+
+  producer_side_service_stub_ = ProducerSideService::NewStub(channel);
+  CHECK(producer_side_service_stub_ != nullptr);
+
+  connect_and_receive_commands_thread_ = std::thread{[this] { ConnectAndReceiveCommandsThread(); }};
+}
+
+void CaptureEventProducer::ShutdownAndWait() {
+  {
+    absl::WriterMutexLock lock{&shutdown_requested_mutex_};
+    CHECK(!shutdown_requested_);
+    shutdown_requested_ = true;
+  }
+
+  {
+    absl::ReaderMutexLock lock{&context_and_stream_mutex_};
+    if (context_ != nullptr) {
+      LOG("Attempting to disconnect from ProducerSideService as exit was requested");
+      context_->TryCancel();
+    }
+  }
+
+  CHECK(connect_and_receive_commands_thread_.joinable());
+  connect_and_receive_commands_thread_.join();
+
+  producer_side_service_stub_.reset();
+  // If producer_side_service_stub_ held the last reference to a gRPC object,
+  // the internal grpc_shutdown will be executed. This can use a detached thread.
+  // From the docs of grpc_shutdown: "The last call to grpc_shutdown will initiate
+  // cleaning up of grpc library internals, which can happen in another thread".
+  // Give that a moment to complete, as otherwise that could lead to a SIGSEGV on exit.
+  std::this_thread::sleep_for(std::chrono::milliseconds{1});
+}
+
+void CaptureEventProducer::OnCaptureStart() { LOG("CaptureEventProducer called OnCaptureStart"); }
+
+void CaptureEventProducer::OnCaptureStop() { LOG("CaptureEventProducer called OnCaptureStop"); }
```

Comment:
these 2 look like a separate interface

Reply:
Discussed offline with Dimitry. `OnCaptureStart`,`OnCaptureStop` shouldn't be separated from `SendCaptureEvents`,`NotifyAllEventsSent` as they basically represent the implementation of a client of `ProducerSideService` as defined in `producer_side_services.proto`.